Add a DecimalRational subclass to print decimal literals

meganly opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Addresses #17648

#### Brief description of what is fixed or changed
Adds a Decimal object called `DecimalRational` that is a subclass of `Rational`, which prints rationals as a decimal expansion in the `latex`, `pretty`, and `str` printers.

#### Other comments
Because most of Rational's methods are hard-coded to use Rational, I had to change many places in the code from `Rational` to `self.__class__` to create a meaningful subclass. This makes the implementation of the Rational class less brittle so that it's easier for future users to extend the Rational class.

It's not always clear how arithmetic and simplification methods should work with DecimalRationals, so I chose something similar to how Floats and Rationals behave (e.g. `DecimalRational('.5') * Rational(1,5) = DecimalRational('.1')`).


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
  * Added decimal literals to print decimals in the output while using exact arithmetic
<!-- END RELEASE NOTES -->

meganly committed:
Create a DecimalRational class

meganly committed:
Print DecimalRationals as decimals in  latex printer

meganly committed:
Modify __abs__ and __neg__ methods to preserve DecimalRational subclass

meganly committed:
Add an __add__ method

meganly committed:
Add a __sub__ method

meganly committed:
Add __mul__ method with compatible latex printing

meganly committed:
Add __div__ and _eval_power methods

meganly committed:
Do not override Rational constructor of integers

meganly committed:
Test __abs__ and __neg__

meganly committed:
Test __add__ and __sub__

meganly committed:
Test __mul__ and __div__

meganly committed:
Fix radsimp for DecimalRational exponents

meganly committed:
Test _eval_power

meganly committed:
Make powsimp preserve DecimalRational exponents

meganly committed:
Print DecimalRationals as decimals in pretty and str printers

meganly committed:
Print DecimalRationals in repr

meganly committed:
Update docstring

asmeurer commented on sympy/core/add.py:
```diff
@@ -140,7 +140,7 @@ def flatten(cls, seq):
                     return [S.NaN], [], None
                 if coeff.is_Number:
-                    coeff += o
+                    coeff = o.__add__(coeff)
```
Why is this change needed?

meganly replied:
So that adding a DecimalRational and a symbol doesn't move the DecimalRational back to a Rational.  See [this](https://github.com/meganly/sympy/blob/17648_DecimalRationalclass/sympy/core/tests/test_arit.py#L1878) test.

asmeurer replied:
Oh I see, it is o + coeff instead of coeff + o. It should work the same to use `+` here instead of `__add__`. 

I guess most of the stuff here has to do with getting this priority stuff to work right, so that when you combine DecimalRational and Rational it stays as DecimalRational.

meganly replied:
op_priority get the priority stuff right so this change isn't needed

asmeurer commented on sympy/core/basic.py:
```diff
@@ -76,6 +76,7 @@ class Basic(with_metaclass(ManagedProperties)):
     is_Number = False
     is_Float = False
     is_Rational = False
+    is_DecimalRational = False
```
Instead of adding this we should just use `isinstance`. The `is_Class` stuff in SymPy is not great. I would personally rather it weren't there at all.

meganly replied:
Fair enough.  I was trying to be consistent with what Rational does, but would you like me to take it out?

asmeurer replied:
Yeah, remove `is_DecimalRational` and replace `x.is_DecimalRational` with `isinstance(x, DecimalRational)`. 

The only potential issue with this is you might get import errors from circular import issues, but if that happens, it usually works to move the import inside the function where `DecimalRational` is used.

meganly replied:
Done.

asmeurer commented on sympy/core/numbers.py:
```diff
@@ -1754,18 +1764,23 @@ def __div__(self, other):
                     return S.ComplexInfinity
                 else:
                     return Rational(self.p, self.q*other.p, igcd(self.p, other.p))
+            elif isinstance(other, DecimalRational):
```
Why do all the arithmetic operations on `Rational` need to know about `DecimalRational`?

meganly replied:
If you change `Rational` in [this](https://github.com/meganly/sympy/blob/17648_DecimalRationalclass/sympy/core/numbers.py#L1705) line to `self.__class__` you run into a problem with `S.Half`.  The only workarounds I saw were to write new methods for `S.Half` or write methods for `DecimalRational`.  I chose the latter so that the conventions`DecimalRational + Rational = DecimalRational` and `Rational + DecimalRational = DecimalRational` would be consistent.  Also to check for perfect squares so that things like `sqrt(.25)` could be evaluated to `sqrt(.5)`, but not rewrite things like `sqrt(.8)` as `2*sqrt(5)/5`.

meganly replied:
I refactored the code to use `op_priority` and a `_classof` method so the arithmetic operations on `Rational` do not need to know about `DecimalRational`

asmeurer commented on sympy/core/numbers.py:
```diff
@@ -2038,6 +2053,154 @@ def as_coeff_Add(self, rational=False):
+    is_DecimalRational = True
+
+    @cacheit
+    def __new__(cls, *args):
```
Is it necessary to duplicate all this code on `DecimalRational`? It would be better if it weren't duplicated. The way it is here, any updates to Rational will need to also be made to DecimalRational. We would also need to make sure all this code is tested.

meganly replied:
I added tests to illustrate why all the changes I made were necessary.  Maybe there is a more elegant way to do arithmetic with DecimalRationals without them getting moved back to Rational so that the tests still pass.

meganly replied:
Giving DecimalRationals a higher operation priority is the more elegant way to do this so that all this code from Rational doesn't need to be duplicated

asmeurer commented:
It's unfortunate if some of these are required to make this work. Ideally, you could just write

```py
class DecimalRational(Rational):
    pass
```

and only modify the printers. It's not great if things outside of DecimalRational and the printers need to know about it, or if DecimalRational has to duplicate code from Rational to work. 

Some of these workarounds may be unavoidable, but I want to understand why they are required.

oscarbenjamin commented:
Rather than using `self.__class__` it would probably be better to have a `_new` method in `Rational` that can be overridden by `DecimalRational`.

Ideally all of the `__add__` etc methods in numbers.py would use `call_highest_priority` like `Expr`:
https://github.com/sympy/sympy/blob/9be6816a5f30c4d326110f5d6c6f7c99dfeb8958/sympy/core/expr.py#L177-L180
Then it should only be necessary to give `DecimalRational` a higher `op_priority` than `Rational`.

I think that `Add/Mul.flatten` defer to `*` and `+` for `Number` so changes shouldn't be needed there if the basic arithmetic operations are working properly.

sylee957 commented:
I'm not still very much convinced about the idea of how 'decimal' can provide useful.
Should it be like, a different-base representation of rational numbers needed? 

I'm open on introducing lots of separate 'number' classes like base-n representations, Gregorian calendar, continued fractions, ...
However, when used alongside with other sympy modules, I think that they should be coerced into the forms that doesn't confuse the computation, rather aggressively converting other numbers into itself.

I'm not sure that it can avoid the issues with solvers, integrals, matrices and much more,

meganly commented:
> I'm not still very much convinced about the idea of how 'decimal' can provide useful.
> Should it be like, a different-base representation of rational numbers is needed?

I think it's useful see decimals printed while doing exact computations.  I develop math education software that uses Sympy on the backend.  It is important to me to have decimal literals to do basic algebra problems without floating point arithmetic errors.  Things like `9.9**2*x-10.1**2*x`. I don't think a different base representation is needed since this is a printing issue.  In addition to printing decimals, it would also be nice to print mixed numbers.  I want to mix decimals, fractions, and mixed numbers in expressions (to test whether students can convert between these different forms) so it seems necessary to subclass Rational.  However, the way Rational is hard-coded makes it difficult to extend the Rational class.  I'm not the only educator who has struggled with [this](https://stackoverflow.com/questions/35871158/how-to-sanely-extend-sympy-rational) so I appreciate all the help.

@oscarbenjamin, I like your idea but don't understand what you mean by a `_new` method.  Is that different from `__new__`?

asmeurer commented:
I'm not sure exactly what @oscarbenjamin is suggesting, but it's probably related to these ideas https://github.com/sympy/sympy/issues/17280

asmeurer commented:
> However, when used alongside with other sympy modules, I think that they should be coerced into the forms that doesn't confuse the computation, rather aggressively converting other numbers.

I think SymPy fundamentally needs to be able to work with different mathematically equivalent ways of expressing the same thing. Even Float(0.5) and Rational(1, 2) are basically the same number, but SymPy needs to be able to deal with both. What I don't think we can guarantee is if we pass some expression to a function like solve() or simplify(), that the specific types will somehow be preserved in the output. Because what might look like a correspondence might not actually be how things are implemented internally. So for instance `solve(DecimalRational("0.1")*x - 1)` might give `Integer(10)` or it might give `DecimalRational(10.0)`. Which it gives is really a function of how the internal algorithm computes the answer (this is a simple example but you can imagine things can get more complicated).

asmeurer commented:
> Ideally all of the __add__ etc methods in numbers.py would use call_highest_priority like Expr:

A good test if this works or not would be to try to implement it for Float. Right now, the priority of Float + Rational is implemented manually (e.g., [here](
https://github.com/sympy/sympy/blob/72e0994c40fefcca3197008843f2b202ddc42280/sympy/core/numbers.py#L1709)).

op_priority isn't a great way of doing dispatching. There are limitations to how far we can get with it.

oscarbenjamin commented:
> op_priority isn't a great way of doing dispatching

No it's not but it does solve one problem which is the potential inconsistency between `a.__add__(b)` and `b.__radd__(a)`.

smichr commented:
I'm not following this too closely but left a [comment](https://github.com/sympy/sympy/issues/17648#issuecomment-588035254 about converting to repeating decimal form (if that is of any use here).

meganly committed:
Refactor methods to use call_highest_priority

meganly committed:
Revert Add/Mul flatten to * and +

meganly committed:
Remove is_DecimalRational and use isinstance

meganly committed:
Clean up _eval_power method

meganly commented:
I refactored the code using @oscarbenjamin's insightful ideas and rebased onto sympy 1.6.  I'll see if I can incorporate the work of @smichr to print repeating decimals nicely.

oscarbenjamin commented:
I think maybe it will be hard to make this work because any algorithm that decouples the numerator and denominator of a Rational will end up converting DecimalRational to Rational. This would happen e.g. if converting to a Poly as is done in functions like factor and cancel.

An easier way might be:

1. Add options for the printer to display Rational in decimal notation
2. Add an option to use nsimplify automatically when sympifying a Python float.

Those options together would probably make a lot of users happier with sympy's handling of "floats".

asmeurer commented:
Using float literals as decimal inputs without preprocessing will always lead to corner-case issues. Too much information is lost once the string data is converted to a binary machine float. However, that isn't really relevant for this issue because even with a dedicated class you will have to parse the input to it in some way. I think `sympify(string_input, rational=True)` is the best way. I suppose the concern is that there's still information loss, basically, there's no way to tell the difference between `sympify('0.1', rational=True)` and `sympify('1/10')`. So the best you can do is to try to intelligently guess which rationals should be printed as decimals. 

The real issue is how to make it so that DecimalRational "stays" as DecimalRational after passing through some operations. Certain operations obviously shouldn't change it to Rational. For example, if you add something to an expression, it shouldn't modify what was already there. Other things it isn't clear on. If something pulls a rational apart and makes new rationals, it would certainly be nice if it stayed as DecimalRational, but I don't think it would be too surprising if it didn't either. I think `cancel`, for example, will make the coefficients all integers, so personally I think that would be fine if it went to Integer. Maybe it can propagate to other operations and cause problems, but I wouldn't try to overthink this. I think the real issues here have to do with Rational being so hard-coded in the core, and in particular things like `S.Half` making it very difficult to have Rational subclasses.

oscarbenjamin commented:
I'm not thinking of an option to sympify but rather a global option that some users can use so that a float is always sympified to Rational using nsimplify e.g.:
```python
0.1 * x   --->   _sympify(0.1) * x   --->   nsimplify(0.1) * x  --->  Rational(1, 10) * x
```
Then internally the `0.1` float becomes `Rational(1, 10)`. The second part is an option to the printers that can be used with `init_printing()` that will print `Rational(1, 10)` as `0.1`.

Then a user can do
```python
from sympy import *
use_decimal_rational()   # sets both settings

x = Symbol('x')
expr = 0.1 * x ** (1/2) + 2.5
#
# Lots of calculations like solve, Poly, integrate, etc that all work fine because everything is Rational
#
pprint(result)  # Coefficients like the 0.1 printed as decimal rather than a fraction
```
I think the issues that many users have with floats is that they don't actually want to use float/Float. It's just that's what Python syntax does and then sympy seems to penalise them for using decimal notation for non-integers like 0.2.

asmeurer commented:
I think an option like that would be bad. nsimplify can only guess what the user actually typed and there would be weird corner cases (or at least they would feel weird to a user who doesn't understand what is going on).

czgdp1807 commented:
@asmeurer @sylee957 Can this PR be concluded sometime in future?

meganly commented:
My pitch for this PR is that I think a DecimalRational class would be useful for Sympy Gamma.  When ``9.9**2-10.1**2`` is entered in Sympy Gamma it returns ``−3.99999999999999``, which is surprising to someone who is not aware of floating point errors.  Wolfram Alpha returns ``-4``.  A DecimalRational class could also return an exact answer for things like ``sqrt(0.5)`` instead of a fractional approximation.

sylee957 commented:
I feel really weird about how this class should be represented with `p` and `q`, and how this number system really stands between the exact number systems (like rational) and truncated number systems (like float, python decimal).

But if you have originally thought of `p`, `q` representation and have exact arithmetic,
But I'd question why not representing like
1. Finite digits (`Decimal([1], [2, 3]) -> 1.23`)
2. Finite digits + recurring digits (`Decimal([1], [2, 3], [4, 5]) -> 1.2345454545...`)

And rather develop addition and multiplication with actual decimals in the core?

sylee957 commented:
I haven't really heard anything clear about repeating decimals in discussions about this.
I just commented above about the alternative representations that actually uses finite digit, because you should consider how the repeating decimals should be represented if you stick to `p/q`

sylee957 commented:
The other unclearness I see is whether are you going to make a class for exact numbers or approximate numbers.
I think that these two number systems are really different if you try to consider examples like adding `100000000` + `.00000001`, and the approximate number system is not really different than reinventing the floating point system to different base, I'm not sure if you want this or not judging by the small examples.

sylee957 commented:
The final comment I have is how do you expect users to extend on this new number system.
I have first thought that this should be an entry for users to contribute algorithms for implementing decimal number theories.
So I have also thought that having more extended representation like `DecimalRational(*, base)` (For implementing repeating decimals up to any base) can be more general for this.

czgdp1807 commented:
> When `9.9**2-10.1**2` is entered in Sympy Gamma it returns `−3.99999999999999`, which is surprising to someone who is not aware of floating point errors. Wolfram Alpha returns `-4`

Just did some experiments,

```python
In [21]: 9.9**2-10.1**2 # Everything of Python
Out[21]: -3.999999999999986

In [22]: S(9.9)**2 - S(10.1)**2 # Pow And Difference algorithms of SymPy but rep
    ...: resentations of 9.9 and 10.1 are from Python
Out[22]: -3.99999999999999

In [23]: (S(99)/10)**2 - (S(101)/10)**2 # SymPy Rationals
Out[23]: -4
```
Now, to give results similar to Python we need to figure out a way to convert from binary representations of decimals used by Python to exact rationals otherwise errors will persist. Wolfram language may parse decimals in a different manner to give exact results. 

One way to avoid floating point errors is to first convert `arg` in `S(arg)` where `arg` is a floating point number to a `str` and then parsing it using the algorithms implemented in SymPy which may give exact results.

```python
In [26]: str(9.9)
Out[26]: '9.9'

In [27]: str(10.1)
Out[27]: '10.1'
```

That way we may escape the internal representations of Python and store numbers in SymPy's way to give exact results. But again it depends on how lossy `str` is but it is a Python feature.

Bottom line is that, it is a representation problem and we can solve it by changing the way SymPy stores numbers.

References

https://docs.python.org/3/tutorial/floatingpoint.html

oscarbenjamin commented:
> My pitch for this PR is that I think a DecimalRational class would be useful for Sympy Gamma.

Both SymPy Live and SymPy Gamma treat `1/2` as a `Rational` although that would be a float in any Python environment. I think it would be reasonable for them to use `nsimplify` for any `Float` input but remember that the user gave input in decimal and then also give output in decimal. Alternatively any float could be replaced with a symbol for the purposes of any calculation but then displayed as a decimal either exact or `nsimplified` at the end.

The significant of `nsimplify` is that it can make an approximation to get a simple exact form from an approximate float:
```
In [21]: nsimplify(0.33333333333)
Out[21]: 1/3

In [22]: Rational(0.33333333333)
Out[22]: 
 6004799503100613
─────────────────
18014398509481984
```

> But I'd question why not representing like
> 
> 1. Finite digits (`Decimal([1], [2, 3]) -> 1.23`)
> 2. Finite digits + recurring digits (`Decimal([1], [2, 3], [4, 5]) -> 1.2345454545...`)
> 
> And rather redevelop addition and multiplication in the core?

It's easer to do all the calculations in standard rational representation if they are intended to be exact. I expect that computing e.g addition of numbers with different recurring digits would end up just converting them to rational and back.

The purpose of decimal arithmetic as provided in Python's decimal module and also similar libraries in other languages is all about having control of rounding. There are some situations, particularly in finance, where precise decimal rounding rules are a legal requirement and for those cases approximating decimal rounding with binary rounding is not acceptable. If you actually want exact arithmetic then I don't see any advantage in using a decimal representation.

meganly commented:
> Now, to give results similar to Wolfram we need to figure out a way to convert from binary representations of decimals used by Python to exact rationals otherwise errors will persist.

> Bottom line is that, it is a representation problem and we can solve it by changing the way SymPy stores numbers.

Exactly.

>  I think it would be reasonable for them to use `nsimplify` for any `Float` input but remember that the user gave input in decimal and then also give output in decimal.

This is essentially the approach of the `DecimalRational` class.  Instead of `nsimplify`, decimals are be passed in as `DecimalRational('9.9')`.  This constructs the decimal as `Rational('9.9')` so calculations are exact.  The class serves to remember the user gave input in decimal and then also gives the output in decimal.  I'm not sure how else you would do this.

> If you actually want exact arithmetic then I don't see any advantage in using a decimal representation.

Agreed, but there are a lot of people who will use decimals and have no idea that they are not doing exact arithmetic.  I don't know who the target Sympy Gamma user is, but part of why Wolfram Alpha is successful is because it is intuitive.  A middle school student can use Wolfram Alpha as calculator to do decimal arithmetic for their math homework and not have to worry about floating point issues.  I'm developing a custom version of Sympy Gamma for K12 math educators where they can enter expressions in MathQuill and Sympy does the math for them.  This enables educators to easily create homework problems and other online content.

sylee957 commented:
What is the state of implementing the repeating decimal as you have stated in the comment #17648
The only thing I'm not comfortable with is that you are using converting it to float to print it to decimal.

I'm sure that python or mpmath itself have some logic to truncate the numbers into base-10 decimal numbers that could be 'supposedly from', like how `0.1` gets printed back as `0.1`, but relying on this logic to print is not comfortable for me as much as you are not comfortable with `nsimplify`.

Also there would be a possibility of users to use float and decimal in the same expression, and we would end up with expression showing both float and decimal, I would strongly suggest that float and decimal should use separate representations even for finite decimals ![image](https://user-images.githubusercontent.com/34944973/105229176-e0d9c900-5ba6-11eb-9688-4872c5155cc0.png)

meganly commented:
Apologies for the delayed reply.  I have implemented a DecimalRational class, but haven't done any work on implementing the repeating decimal.  Pardon my ignorance, but how do you print rational numbers as decimals without using float?  I think the idea of Sympy Gamma is be intuitive so Floats and DecimalRationals should both print as a decimal.

oscarbenjamin commented:
> how do you print rational numbers as decimals without using float?

The idea is to extract one digit at a time using divmod. Something like:
```python
In [23]: q = Rational(1, 7)

In [24]: for _ in range(10):
    ...:     d, q = divmod(q*10, 10)
    ...:     print(d)
    ...: 
0
1
4
2
8
5
7
1
4
2
```

sylee957 commented:
There is an example implementation in https://stackoverflow.com/questions/249372/how-to-calculate-recurring-digits
For finding the integer part of the decimal, we can avoid extracting digits one by one, but just call `divmod` once to split the rational into 'integer part' and the 'proper fraction' (where p < q)
The important part of building the algorithm is that we need to keep track of the repeating occurrence of 'remainder' than the 'quotient' or you can fall into the some number theoretic trap since there are lots of examples that have more than 10 repeating digits https://en.wikipedia.org/wiki/Repeating_decimal
